_posts/_2021-08-31-building-opensearch-1.1-distributions.markdown

```diff
@@ -0,0 +1,44 @@
+---
+layout: post
+title:  "Building OpenSearch 1.1 Distributions"
+authors: 
+  - dblock
+date: 2021-08-31
+categories: 
+  - technical-post
+twittercard:
+  description: "This post details how we produce reliable and repeatable distributions of OpenSearch."
+---
+The OpenDistro for Elasticsearch (ODFE) was essentially a package containing Elasticsearch OSS and a dozen open-source plugins. Its release process began by picking up a stable version of Elasticsearch OSS, incrementing all the plugin version numbers, and going through a development cycle. Plugins went from `-SNAPSHOT` to `-beta`, then `-rc`, then finally to a release.
+
+This worked well until the [OpenSearch fork](https://aws.amazon.com/blogs/opensource/introducing-opensearch/). It was decided that OpenSearch would be a product with two distributions at launch: a full distribution that includes a dozen plugins, and a minimal distribution without any plugins. This made sense because most customers either wanted a well tested distribution with everything, or are technically savvy enough to assemble their favorite flavor of all the things. 
+
+The full distribution of OpenSearch was similar to ODFE, but with one major difference - the OpenSearch service would be developed in parallel with the plugins. The complete OpenSearch distribution became effectively a monolith spanning multiple GitHub repositories. You just can't both develop all the moving parts simultaneously, and have stable dependencies of [OpenSearch](https://github.com/opensearch-project/OpenSearch), [common-utils](https://github.com/opensearch-project/common-utils), [job-scheduler](https://github.com/opensearch-project/job-scheduler), to name a few, at the same time.
+
+To produce a reliable monolith that can be trusted we decided to build the entire distribution from source. We developed a [manifest-based workflow](https://github.com/opensearch-project/opensearch-build/tree/main/bundle-workflow) to accomplish this, which consists of the following steps.
+
+1. Take an input manifest ([example](https://github.com/opensearch-project/opensearch-build/blob/main/manifests/opensearch-1.1.0.yml)), build all components from source, and generate a build manifest. Today we build x64 and arm64 flavors.
+2. Take a build manifest ([example](https://github.com/opensearch-project/opensearch-build/blob/main/bundle-workflow/tests/tests_manifests/data/opensearch-build-1.1.0.yml)) and assemble a distribution.
+3. Sign the artifacts.
+4. Run integration, backwards-compatibility, performance, and upgrade tests against the assembled distribution.
+5. Publish maven artifacts to our snapshot maven repository and the distributions to S3.
+
+Here's all the inputs and outputs for OpenSearch 1.1.
+
+* Maven releases are staged in the [AWS OSS Maven Repository](https://aws.oss.sonatype.org/content/repositories/releases/org/opensearch/) before being promoted to Maven Central. 
```

Comment:
This URL is the last step at release time, its not used for staging. The flow is as follows:

A nightly job runs that runs three separate jobs in parallel.
1. A snapshot build that builds and pushes -SNAPSHOT maven artifacts to https://aws.oss.sonatype.org/content/repositories/snapshots/
2&3. arm64/x64 builds.  Each of these produce the bundle, plugin zips, and maven artifacts and pushes them to S3.

When this job completes the test step is triggered.  On a successful test pass we promote those artifacts to a staging S3 bucket as the latest 'good' build.

On or before release day we run the 'release workflow' giving a build ID as input.  That release workflow picks up the artifacts, signs them and publishes.  This means bundles & zips are signed and placed in artifacts.opensearch.org/releases and maven artifacts are staged & released to https://aws.oss.sonatype.org/content/repositories/releases/org/opensearch/.  That is then mirrored to maven central.